cl/donate/api_views.py

```diff
@@ -0,0 +1,215 @@
+
+        return user
+
+    def _get_membership_data(
+        self, webhook_data: dict[str, str]
+    ) -> dict[str, str]:
+        """
+        Extracts relevant membership information from a Neon webhook payload.
+
+        Args:
+            webhook_data (dict[str, str]): The payload received from a Neon
+            webhook.
+
+        Returns:
+            dict[str, str]: A dictionary containing the extracted membership
+            data, with the following keys:
+            - membershipId: Unique identifier for the membership.
+            - accountId: Identifier for the user account associated with the
+            membership.
+            - termEndDate: The date the membership is scheduled to terminate
+            - membershipName: The name of the membership level the user is
+            enrolled in.
+        """
+        if "membership" in webhook_data["data"]:
+            data = webhook_data["data"]["membership"]
+            membership = {
+                "membershipId": data["membershipId"],
+                "membershipName": data["membershipName"],
+            }
+        else:
+            data = webhook_data["data"]
+            membership = {
+                "membershipId": data["id"],
+                "membershipName": data["membershipLevel"]["name"],
+            }
+
+        membership.update(
+            {
+                "accountId": data["accountId"],
+                "termEndDate": data["termEndDate"],
+            }
+        )
+
+        return membership
+
+    def _map_trigger_value(self, trigger_event: str) -> int:
+        """
+        Maps a string trigger event received from a Neon webhook to the
+        corresponding integer value representing the trigger event type in
+        the NeonWebhookEvents model.
+
+        Args:
+            trigger_event (str): The string representing the trigger event
+            type received in the Neon webhook payload.
+
+        Returns:
+            int: The integer value of the trigger event type according to
+            the NeonWebhookEvents model
+        """
+        match trigger_event:
+            case "createMembership":
+                trigger = NeonWebhookEvents.MEMBERSHIP_CREATION
+            case "updateMembership":
+                trigger = NeonWebhookEvents.MEMBERSHIP_UPDATE
+            case "editMembership":
+                trigger = NeonWebhookEvents.MEMBERSHIP_EDIT
+            case "deleteMembership":
+                trigger = NeonWebhookEvents.MEMBERSHIP_DELETE
+
+        return trigger
+
+    def _store_webhook_payload(self, webhook_data) -> None:
+        trigger = self._map_trigger_value(webhook_data["eventTrigger"])
+        if trigger != NeonWebhookEvents.MEMBERSHIP_DELETE:
+            membership_data = self._get_membership_data(webhook_data)
+        else:
+            membership_data = webhook_data["data"]["membership"]
+        NeonWebhookEvents.objects.create(
+            content=json.dumps(webhook_data, default=str),
```

Comment:
I think you'll want indent=2 here, no?

Reply:
Thank you for bringing this to my attention.

I used the `dumps` method to address an issue with `datetime` objects because they are not directly JSON serializable and I was getting an error when the code tried to store the payload. This method's `default` argument provides a function that gets called for objects that can't be serialized. In our case, It just converts everything it doesn't know to strings. This approach is fine for serialization but the output is not a human-readable JSON representation (the dumps method adds a bunch of new lines and backslashes to the string).

While reviewing your comments, I realize we can handle `datetime` objects without relying on the `dumps` method. We can pass an `encoder` subclass to the Django [JSONField](https://docs.djangoproject.com/en/5.0/ref/models/fields/#jsonfield) to serialize data types not supported by the standard JSON serializer and Django provides the handy [DjangoJSONEncoder](https://docs.djangoproject.com/en/5.0/topics/serialization/#djangojsonencoder) class, specifically designed to handle `datetime`, `date`, `time`, and `UUID` objects effortlessly.

We just need to tweak the content field in the `NeonWebhookEvent` model like this:

```python
from django.core.serializers.json import DjangoJSONEncoder

class NeonWebhookEvent(AbstractDateTimeModel):
    ...
    content = models.JSONField(  # type: ignore
        encoder=DjangoJSONEncoder,
        help_text="The content of the payload of the POST request.",
    )

```

Reply:
Well that's very handy.